src/crystal/system/unix/file.cr

```diff
@@ -105,6 +105,28 @@ module Crystal::System::File
     ret
   end
 
+  def self.readlink(path) : String
+    buf = Bytes.new 256
+    # First pass at 256 bytes handles all normal occurences in 1 system call.
+    # Second pass at 1024 bytes handles outliers?
+    # Third pass is the max or double what Linux/MacOS can store.
+    3.times do |iter|
+      bytesize = LibC.readlink(path, buf, buf.bytesize)
+      if bytesize == -1
+        raise Errno.new("readlink")
+      elsif bytesize == buf.bytesize
+        break if iter >= 2
+        buf = Bytes.new(buf.bytesize * 4)
+      else
+        buf[bytesize] = 0_u8
+        return String.new(buf[0, bytesize])
```

Comment:
This would be an alternative, but there is no real difference between both:

```suggestion
        return String.new(buf.to_unsafe, bytesize)
```

`buf[bytesize] = 0_u8` is unneccessary, though. The `String` constructor ensures the string ends with a zero byte. And the C string should be zero delimited anyway.

Reply:
`= 0_u8` was removed.

From the linux man page:

> Conforming applications should not assume that the returned contents of the symbolic link are null-terminated.